Fix dictwatcherdel crash

blueyed opened the pull request:
Fixes https://github.com/neovim/neovim/issues/11188.

blueyed committed:
Test and initial fix for crash with dictwatcherdel

Fixes https://github.com/neovim/neovim/issues/11188.

blueyed commented:
Hmm, appears to still crash on OpenBSD:
```
[ RUN      ] VimL dictionary notifications does not crash when using dictwatcherdel in callback: -- Tests exited non-zero: SIGPIPE
-- No output to stderr.
CMake Error at /home/build/neovim/cmake/RunTests.cmake:85 (message):
  functional tests failed with error: SIGPIPE
```

blueyed commented on src/nvim/eval/typval.c:
```diff
@@ -1139,7 +1140,11 @@ bool tv_dict_watcher_remove(dict_T *const dict, const char *const key_pattern,
   }
 
   QUEUE_REMOVE(w);
```
Does `QUEUE_REMOVE` need to be delayed also?

justinmk commented on src/nvim/eval/typval.c:
```diff
@@ -1208,6 +1213,9 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
       watcher->busy = false;
+      if (watcher->needs_free) {
+        tv_dict_watcher_free(watcher);
+      }
```
Does this mean `dictwatcherdel` would need to wait for the notification to trigger again, before the object is actually freed? I guess that could be surprising for someone watching memory usage after calling `dictwatcherdel`.

bfredl replied:
I don't see the problem, is this not synchronous already? To me it looks like this free call will free the dict watcher in the _same_ event as the `callback_call` which caused it.

blueyed replied:
Yes.

erw7 commented:
> Any ideas about the failiures still?

`w(QUEUE *)` exists within the `watcher(DictWatcher *w)`. Thus, if `tv_dict_watcher_free()` is called in `QUEUE_FOREACH`, `q` refers to the freed memory when evaluating `(q)! = (h)`. You will probably need to change it to:

```C
--- a/src/nvim/eval/typval.c
+++ b/src/nvim/eval/typval.c
@@ -1139,10 +1139,10 @@ bool tv_dict_watcher_remove(dict_T *const dict, const char *const key_pattern,
     return false;
   }

-  QUEUE_REMOVE(w);
   if (watcher->busy) {
     watcher->needs_free = true;
   } else {
+    QUEUE_REMOVE(w);
     tv_dict_watcher_free(watcher);
   }
   return true;
@@ -1205,8 +1205,9 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,

   typval_T rettv;

-  QUEUE *w;
-  QUEUE_FOREACH(w, &dict->watchers) {
+  QUEUE *w = QUEUE_HEAD(&dict->watchers);
+  while(1) {
+    QUEUE *next = w->next;
     DictWatcher *watcher = tv_dict_watcher_node_data(w);
     if (!watcher->busy && tv_dict_watcher_matches(watcher, key)) {
       rettv = TV_INITIAL_VALUE;
@@ -1214,10 +1215,15 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
       callback_call(&watcher->callback, 3, argv, &rettv);
       watcher->busy = false;
       if (watcher->needs_free) {
+        QUEUE_REMOVE(w);
         tv_dict_watcher_free(watcher);
       }
       tv_clear(&rettv);
     }
+    if (next == &dict->watchers) {
+      break;
+    }
+    w = next;
   }

   for (size_t i = 1; i < ARRAY_SIZE(argv); i++) {
```

I think there is no problem in calling `QUEUE_REMOVE (w)` with `tv_dict_watcher_remove()`.

erw7 committed:
Change QUEUE_FOREACH macro to use while instead of for

blueyed commented:
@erw7 Thanks!

Should we rewrite the `QUEUE_FOREACH` macro like that?

erw7 commented:
My patch can be simpler.

```C
--- a/src/nvim/eval/typval.c
+++ b/src/nvim/eval/typval.c
@@ -1205,8 +1205,9 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,

   typval_T rettv;

-  QUEUE *w;
-  QUEUE_FOREACH(w, &dict->watchers) {
+  QUEUE *w = QUEUE_HEAD(&dict->watchers);
+  while(w != &dict->watchers) {
+    QUEUE *next = w->next;
     DictWatcher *watcher = tv_dict_watcher_node_data(w);
     if (!watcher->busy && tv_dict_watcher_matches(watcher, key)) {
       rettv = TV_INITIAL_VALUE;
@@ -1218,6 +1219,7 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
       }
       tv_clear(&rettv);
     }
+    w = next;
   }

   for (size_t i = 1; i < ARRAY_SIZE(argv); i++) {
```

erw7 commented:
> Should we rewrite the QUEUE_FOREACH macro like that? Not possible due to the inner while loop though, is it?

Macros can be defined as follows:

```C
--- a/src/nvim/lib/queue.h
+++ b/src/nvim/lib/queue.h
@@ -39,6 +39,15 @@ typedef struct _queue {
   for (  /* NOLINT(readability/braces) */ \
       (q) = (h)->next; (q) != (h); (q) = (q)->next)

+#define QUEUE_WHILE(q, h, code) \
+  (q) = (h)->next; \
+  while((q) != (h)) { \
+    QUEUE *next = q->next; \
+    code \
+    (q) = next; \
+  }
+
+
 // ffi.cdef is unable to swallow `bool` in place of `int` here.
 static inline int QUEUE_EMPTY(const QUEUE *const q)
   FUNC_ATTR_ALWAYS_INLINE FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT
```

Other than this, there are two places where `while` is used to release memory in the loop. If we define a `QUEUE_WHILE` macro like the one above, I feel like it can be replaced.

https://github.com/neovim/neovim/blob/c5c06665ed302e774dc1a01bed949815b4b7c8c3/src/nvim/event/multiqueue.c#L122-L131

https://github.com/neovim/neovim/blob/c5c06665ed302e774dc1a01bed949815b4b7c8c3/src/nvim/os/pty_process_win.c#L297-L309

blueyed commented:
Awesome.
Couldn't this replace `QUEUE_FOREACH` then?  (to make it safe by default)
I assume there is no more overhead in practice, is there?

erw7 commented:
> Couldn't this replace QUEUE_FOREACH then?

I have only confirmed that the compilation passes, but we can replace the `QUEUE_FOREACH` macro as follows:

https://github.com/blueyed/neovim/compare/fix-dictwatcherdel-crash...erw7:fix-dictwatcherdel-crash

> I assume there is no more overhead in practice, is there?

There should probably not be enough overhead to be a problem.

blueyed commented:
@erw7 
I think your changes make sense.  Do you want to create a PR for it, or should I just pull them in here?

erw7 commented:
> Update: merged your branch here.

@blueyed thanks!

blueyed commented:
Reverted needs_free, not necessary anymore.

btw: I've noticed an issue with the syntax highlighting for the following, which marks the outer "{" as `cErrInParen` (cErrInParen cBadBlock cParen cBlock), and still another when using `:let g:c_no_curly_error = 1` (https://github.com/vim/vim/issues/3551).

justinmk commented:
> btw: I've noticed an issue with the syntax highlighting for the following, which marks the outer "{" as `cErrInParen` (cErrInParen cBadBlock cParen cBlock), and still another when using `:let g:c_no_curly_error = 1` ([vim/vim#3551](https://github.com/vim/vim/issues/3551)).

Yep, known issue. Not really worth our time to fix it, if treesitter works out better.

erw7 commented on src/nvim/eval/typval.c:
```diff
@@ -1201,16 +1201,15 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
       watcher->busy = true;
       callback_call(&watcher->callback, 3, argv, &rettv);
       watcher->busy = false;
-      tv_clear(&rettv);
```
Oops! `tv_clear(&rettv)`; seems to be leaking in my commit https://github.com/neovim/neovim/pull/11189/commits/54ce37f5698ce6471adf9b373d1dd146c65b29dc.

blueyed replied:
Added in a4436ba - is that correct?

erw7 replied:
Yes, that ’s fine.

erw7 commented:
> Reverted needs_free, not necessary anymore.

We still need `needs_free`. If you release the node in the callback, `watcher` does not point to valid memory at `watcher->busy = false;`.

https://github.com/neovim/neovim/blob/df31200a6c8a87df092c4895e02fa65486d819ac/src/nvim/eval/typval.c#L1209-L1210

blueyed commented:
> > Reverted needs_free, not necessary anymore.
> 
> We still need `needs_free`. If you release the node in the callback, `watcher` does not point to valid memory at `watcher->busy = false;`.
> 
> https://github.com/neovim/neovim/blob/df31200a6c8a87df092c4895e02fa65486d819ac/src/nvim/eval/typval.c#L1209-L1210

Ok, reverted.  I've only tested it using the test I've added.

blueyed commented:
> Nice! Is this ready then? Would be good to squash such that @erw7 is author of the relevant commit.

Will do.

blueyed commented:
Squashed into two commits.

blueyed committed:
tv_dict_watcher_remove: delay freeing with busy queue

blueyed commented:
There's still a problem here though (need to come up with a test): `tv_dict_watcher_remove` might free a non-current entry from the queue in https://github.com/blueyed/neovim/blob/4efb68605f61cc59a260be4e31a9900f3fddee13/src/nvim/eval/typval.c#L1146 still.
Pushed an attempt to address this: `tv_dict_watcher_remove` detects if the queue is busy altogether if any watcher is busy (might be better to set this on the dict maybe?), and then `tv_dict_watcher_notify` loops over the whole queue afterwards for removing/freeing.

blueyed committed:
test

blueyed committed:
tv_dict_watcher_notify: any_needs_free

erw7 commented:
@blueyed Please pull https://github.com/erw7/neovim/commit/e82006805a215c28c9519b258260c410b7a0b6d4. Because of my patch, when a callback(`W2`) was registered in the callback(`W1`) in the following cases, the callback(`W2`) was not called.

```vim
let g:d={}

function! W1(...) abort
  call dictwatcheradd(g:d, '*', function('W2'))
endfunction

function! W2(...) abort
  echomsg 'W2 was called'
endfunction

call dictwatcheradd(g:d, '*', function('W1'))
let g:d.foo = 123
```

erw7 commented:
Or since it is no longer necessary with https://github.com/neovim/neovim/pull/11189/commits/6e6af2779ff0d31b74a1c2684c4d798480abab92, it may be better to revert https://github.com/neovim/neovim/pull/11189/commits/4efb68605f61cc59a260be4e31a9900f3fddee13.

blueyed commented:
> it may be better to revert [4efb686](https://github.com/neovim/neovim/commit/4efb68605f61cc59a260be4e31a9900f3fddee13).

I think it is a good change by itself, and would rather keep it.

I am not sure if a watcher added in the active callback should be called already?  It was not registered when the dict was changed after all.  (6e6af27 does not change that, i.e. your example still does echo nothing).

Should be added as a test when we decide how it should behave.

justinmk commented:
is `any_needs_free` just a perf optimization? I would suggest removing that, because (1) watchers list is usually very small, and (2) it reduces "coverage" of that codepath.

erw7 commented:
> I am not sure if a watcher added in the active callback should be called already?

I have no use cases, but the behavior of this branch is confusing:

```vim
let g:d = {}
let g:a = []

function! W1(...) abort
  call dictwatcheradd(g:d, '*', function('W4'))
endfunction

function! W2(...) abort
endfunction

function! W3(...) abort
  call dictwatcheradd(g:d, '*', function('W5'))
endfunction

function! W4(...) abort
  call add(g:a, 'W4 was called')
endfunction

function! W5(...) abort
  call add(g:a, 'W5 was called')
endfunction

call dictwatcheradd(g:d, '*', function('W1'))
let g:d.foo = 123  " this branch: g:a == [], master: g:a == ['W4 was called']

unlet g:d
let g:d = {}
let g:a = []
call dictwatcheradd(g:d, '*', function('W1'))
call dictwatcheradd(g:d, '*', function('W2'))
let g:d.foo = 123  "this branch and master: g:a == ['W4 was called']

unlet g:d
let g:d = {}
let g:a = []
call dictwatcheradd(g:d, '*', function('W1'))
call dictwatcheradd(g:d, '*', function('W3'))
let g:d.foo = 123  "this branch and master: g:a == ['W4 was called', 'W5 was called']

unlet g:d
let g:d = {}
let g:a = []
call dictwatcheradd(g:d, '*', function('W2'))
call dictwatcheradd(g:d, '*', function('W3'))
let g:d.foo = 123  "this branch: g:a == [], master: g:a == ['W5 was called']
```

I think that it is better to make the same behavior as the master by modifying it as follows(https://github.com/erw7/neovim/commit/e82006805a215c28c9519b258260c410b7a0b6d4).

```diff
diff --git a/src/nvim/eval/typval.c b/src/nvim/eval/typval.c
index e90e4117a..36875bb9d 100644
--- a/src/nvim/eval/typval.c
+++ b/src/nvim/eval/typval.c
@@ -1211,7 +1211,7 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,

   bool any_needs_free = false;
   QUEUE *w;
-  QUEUE_FOREACH(w, &dict->watchers, {
+  for(w = (&dict->watchers)->next; w != &dict->watchers; w = w->next) {
     DictWatcher *watcher = tv_dict_watcher_node_data(w);
     if (!watcher->busy && tv_dict_watcher_matches(watcher, key)) {
       rettv = TV_INITIAL_VALUE;
@@ -1223,7 +1223,7 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
         any_needs_free = true;
       }
     }
-  })
+  }
   if (any_needs_free) {
     QUEUE_FOREACH(w, &dict->watchers, {
       DictWatcher *watcher = tv_dict_watcher_node_data(w);
```

erw7 commented:
> Should be added as a test when we decide how it should behave.

I think that this branch can be used as it is, but it is better to create a test for the following cases to clarify the behavior.

```vim
let g:d = {}
let g:a = []

function! W1(...) abort
  call dictwatcherdel(g:d, '*', function('W2'))
endfunction

function! W2(...) abort
  call add(g:a, 'W2 was called')
endfunction

call dictwatcheradd(g:d, '*', function('W1'))
call dictwatcheradd(g:d, '*', function('W2'))
let g:d.foo = 123  " this branch: g:a == ['W2 was called'], master: g:a == []

unlet g:d
let g:d = {}
call dictwatcheradd(g:d, '*', function('W2'))
call dictwatcheradd(g:d, '*', function('W1'))
let g:d.foo = 123  " this branch and master: g:a == ['W2 was called']
```

justinmk commented:
> I am not sure if a watcher added in the active callback should be called already? It was not registered when the dict was changed after all. ([6e6af27](https://github.com/neovim/neovim/commit/6e6af2779ff0d31b74a1c2684c4d798480abab92) does not change that, i.e. your example still does echo nothing).

Whatever results in the simplest implementation :)

blueyed commented:
> is `any_needs_free` just a perf optimization? I would suggest removing that, because (1) watchers list is usually very small, and (2) it reduces "coverage" of that codepath.

Yes, for optimization.  I think it makes sense also with small lists in general, and we should not sacrifize anything for coverage purposes, but rather cover it then (assuming it only handles one branch there currently).

blueyed committed:
more tests

justinmk commented:
> Actually both branches are covered:

I don't doubt it, but  removing it makes it "more covered". And the optimization is almost certainly negligible.

blueyed commented:
> I don't doubt it, but removing it makes it "more covered".

No.  You still have the macro there (which is only partially covered for (to me) unknown reasons.
Actually having the extra code (that is covered) increases the percentage of coverage.
Also, the (first line of the) macro (which is only partially covered in other places) is covered there?!  https://codecov.io/gh/neovim/neovim/compare/ab860cb5f6c749b937e0986167702a36e45000ca...5a7d127721b7054ef97439d54d4ba648051f9f0f/diff#D5-1229...1234

I'd not pay too much attention to the details with regard to partial branch coverage with the macro itself, and would also still keep it if it would show up wrongly.  In this case however we have all branches covered and the (minor) optimization.

I'm still not sure about the `queue_is_busy` detection, and if it should have a flag on the queue (or dict?) instead.

justinmk commented:
> I'm still not sure about the `queue_is_busy` detection, and if it should have a flag on the queue (or dict?) instead.

If this is asking about behavior my suggestion is still to just do whatever is easiest to implement.

blueyed commented:
Yeah, then that's that (since it there already).
But it detects it by checking if any of the watchers is busy, which kind of feel like a hack to me (and only works for when it comes from there - have not really checked if that's the only code path).

Given that I think this is ready for now.

blueyed commented:
Can merge it manually tomorrow, squashing it into two commits already then I think.

teto commented:
@blueyed feel free to rebase/squash, would be nice to get it in 0.5

janlazo commented:
Fails on Windows builds. https://github.com/neovim/neovim/actions/runs/703459630

erw7 commented:
> Fails on Windows builds. https://github.com/neovim/neovim/actions/runs/703459630

I think you can build it if you make the following changes.

```diff
diff --git a/src/nvim/os/pty_process_win.c b/src/nvim/os/pty_process_win.c
index 502b7ffa5..94444e4d2 100644
--- a/src/nvim/os/pty_process_win.c
+++ b/src/nvim/os/pty_process_win.c
@@ -505,11 +505,11 @@ static int build_env_block(dict_T *denv, wchar_t **env_block)
   *env_block = xmalloc(sizeof(**env_block) * env_block_len);
   wchar_t *pos = *env_block;

-  QUEUE_FOREACH(q, &env_q) {
+  QUEUE_FOREACH(q, &env_q, {
     EnvNode *env_node = QUEUE_DATA(q, EnvNode, node);
     memcpy(pos, env_node->str, env_node->len * sizeof(*pos));
     pos += env_node->len;
-  }
+  })

   *pos = L'\0';
```

janlazo commented:
@erw7 It works. Thank you.

Closing in favor of https://github.com/neovim/neovim/pull/14259
